inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java

```diff
@@ -126,9 +138,48 @@ public boolean isConnected()
                 }
             }
             if (hit.getHighlight() != null) {
+
+                // Highlights from elastic search are small sections of the document text
+                // with the keywords surrounded by the <em> tags. There are no offset information
+                // for the highlights or the keywords in the document text. There is a feature
+                // request for it (https://github.com/elastic/elasticsearch/issues/5736).
+                // Until this feature is implemented, we currently try to find the keywords offsets
+                // by finding the matching highlight in the document text, then the keywords offset
+                // within highlight using <em> tags.
+                String originalText = hit.get_source().getDoc().getText();
+
                 // There are highlights, set them in the result
-                List<String> highlights = new ArrayList<>();
-                highlights.add(hit.getHighlight().getDoctext().get(0));
+                List<ExternalSearchHighlight> highlights = new ArrayList<>();
+                for (String highlight : hit.getHighlight().getDoctext()) {
+
+                    // remove markers from the highlight
+                    String highlight_clean = highlight.replace(highlight_start_tag, "")
+                        .replace(highlight_end_tag, "");
+
+                    // find the matching highlight offset in the original text
+                    int highlight_start_index = originalText.indexOf(highlight_clean);
+
+                    // find offset to all keywords in the highlight
+                    // they are enclosed in <em> </em> tags in the highlight
+                    String highlightTemp = highlight;
+                    ArrayList<OffsetSpan> offsets = new ArrayList<>();
+                    while (highlightTemp.contains(highlight_start_tag)) {
+                        int start = highlight_start_index +
+                            highlightTemp.indexOf(highlight_start_tag);
+                        highlightTemp = highlightTemp.replaceFirst(highlight_start_tag, "");
+                        int end = highlight_start_index +
+                            highlightTemp.indexOf(highlight_end_tag);
+                        highlightTemp = highlightTemp.replaceFirst(highlight_end_tag, "");
+                        offsets.add(new OffsetSpan(start, end));
+                    }
+
+                    try {
+                        highlights.add(new ExternalSearchHighlight(highlight, offsets));
+                    }
+                    catch (IOException e) {
```

Comment:
It would be cleaner and clearer there to explicitly check for empy `offsets` and log appropriately rather than depend on the validation of `ExternalSearchHighlight` object

Reply:
Added validation to check empty `offsets`.